tools/shared/src/main/scala/org/scalajs/core/tools/classpath/PartialClasspath.scala

```diff
@@ -56,23 +59,78 @@ final class PartialClasspath(
   }
 
   /** Constructs an ordered list of JS libraries to include. Fails if:
+   *  - Resource names do not identify a unique resource on the classpath
    *  - Dependencies have cycles
    *  - Not all dependencies are available
    */
   protected def resolveDependencies(
       filter: DependencyFilter): List[ResolvedJSDependency] = {
-    val flatDeps = filter(dependencies.flatMap(_.flatten))
+    val resourceNames = collectAllResourceNames()
+    val resolvedJSLibs = resolveAllResourceNames(resourceNames)
+
+    val allFlatDeps = for {
+      manifest <- dependencies
+      dep <- manifest.libDeps
+    } yield {
+      new FlatJSDependency(manifest.origin, resolvedJSLibs(dep.resourceName),
+          dep.dependencies.map(resolvedJSLibs), dep.commonJSName)
+    }
+
+    val flatDeps = filter(allFlatDeps)
     val includeList = JSDependencyManifest.createIncludeList(flatDeps)
 
-    val missingDeps = includeList.filterNot { info =>
-      availableLibs.contains(info.resourceName)
+    for (info <- includeList)
+      yield new ResolvedJSDependency(availableLibs(info.relPath), info)
+  }
+
+  /** Collects all the resource names mentioned in the manifests.
+   *  @return Map from resource name to the list of origins mentioning them
+   */
+  private def collectAllResourceNames(): Map[String, List[Origin]] = {
+    val nameOriginPairs = for {
+      manifest <- dependencies.toList
+      dep <- manifest.libDeps
+      resourceName <- dep.resourceName :: dep.dependencies
+    } yield (resourceName, manifest.origin)
+
+    nameOriginPairs.groupBy(_._1).mapValues(_.map(_._2))
+  }
+
+  /** Resolves all the resource names wrt to the current classpath.
+   *  @throws JSLibResolveException if any of the resource names cannot be resolved
+   *  @return Map from resource name to relative path
+   */
+  private def resolveAllResourceNames(
+      allResourceNames: Map[String, List[Origin]]): Map[String, String] = {
+    val problems = mutable.ListBuffer.empty[Problem]
+    val resolvedLibs = Map.newBuilder[String, String]
+
+    for ((resourceName, origins) <- allResourceNames) {
+      resolveResourceName(resourceName, origins).fold[Unit](
+          problems += _,
+          resolvedLibs += resourceName -> _)
     }
 
-    if (missingDeps.nonEmpty)
-      throw new MissingJSLibException(missingDeps)
+    if (problems.nonEmpty)
+      throw new JSLibResolveException(problems.toList)
 
-    for (info <- includeList)
-      yield new ResolvedJSDependency(availableLibs(info.resourceName), info)
+    resolvedLibs.result()
```

Comment:
Ah that's what you meant. Well I don't have any strong feeling either way in this case, in the sense that the result of this method can be thought of as being either a map or throwing an exception. But I'm pretty sure that so far, the throws have not been accompanied by an `else`.

I'll change it in the morning.